src/diff/children.js

```diff
@@ -56,162 +58,196 @@ export function diffChildren(
 		}
 	}
 
-	i = 0;
-	newParentVNode._children = toChildArray(
-		newParentVNode._children,
-		childVNode => {
-			if (childVNode != null) {
-				childVNode._parent = newParentVNode;
-				childVNode._depth = newParentVNode._depth + 1;
+	newParentVNode._children = [];
+	for (i = 0; i < renderResult.length; i++) {
+		childVNode = renderResult[i];
 
-				// Check if we find a corresponding element in oldChildren.
-				// If found, delete the array item by setting to `undefined`.
-				// We use `undefined`, as `null` is reserved for empty placeholders
-				// (holes).
-				oldVNode = oldChildren[i];
+		if (childVNode == null || typeof childVNode == 'boolean') {
+			childVNode = newParentVNode._children[i] = null;
+		}
+		// If this newVNode is being reused (e.g. <div>{reuse}{reuse}</div>) in the same diff,
+		// or we are rendering a component (e.g. setState) copy the oldVNodes so it can have
+		// it's own DOM & etc. pointers
+		else if (typeof childVNode == 'string' || typeof childVNode == 'number') {
+			childVNode = newParentVNode._children[i] = createVNode(
+				null,
+				childVNode,
+				null,
+				null,
+				childVNode
+			);
+		} else if (Array.isArray(childVNode)) {
```

Comment:
Putting Array.isArray into a utility function added a byte. Using a `.pop` property check saves us some bytes (~10) but I worry that the property access will quickly de-opt and become megamorphic since it'll sometimes be a VNode and sometimes by an array. So I think imma leave this for now. Can revisit later

Reply:
Good point! iirc checking for `.pop()` is much slower than `Array.isArray()` 👍

Reply:
Yep. It used to be the other way around. Not sure about other engines.  

https://esbench.com/bench/57f1327e330ab09900a1a1d5

Reply:
![](https://i.hrmny.sh/R0gCVArsPZ.png)

firefox